c10/cuda/CUDAException.cpp

```diff
@@ -26,10 +111,197 @@ void c10_cuda_check_implementation(
+  // This is likely to be the longest-lasting hold on the mutex, but
+  // we only expect it to be called in cases where we're already failing
+  // and speed is no longer important
+  const std::lock_guard<std::mutex> lock(read_write_mutex);
+
+  std::vector<DeviceAssertionsData> device_assertions_data;
+  for (const auto& x : uvm_assertions) {
+    if (x) {
+      device_assertions_data.push_back(*x);
+    } else {
+      device_assertions_data.emplace_back();
+    }
+  }
+
+  return std::make_pair(device_assertions_data, kernel_launches);
+}
+
+DeviceAssertionsData* CUDAKernelLaunchRegistry::
+    get_uvm_assertions_ptr_for_current_device() {
+  // Need a lock here so there's not race-condition on creating the new device
+  // assertions buffer
+  const std::lock_guard<std::mutex> lock(read_write_mutex);
+
+  if (!do_all_devices_support_managed_memory) {
+    return nullptr;
+  }
+
+  const auto device_num = get_device_id();
+
+  // If we've already set up this GPU with managed memory, return a pointer to
+  // the managed memory
+  if (uvm_assertions.at(device_num)) {
+    return uvm_assertions.at(device_num).get();
+  }
+
+  // Otherwise, set up the GPU to be able to use the device-side assertion
+  // system
+  DeviceAssertionsData* uvm_assertions_ptr = nullptr;
+
+#ifdef TORCH_USE_CUDA_DSA
+  C10_CUDA_ERROR_HANDLED(
+      cudaMallocManaged(&uvm_assertions_ptr, sizeof(DeviceAssertionsData)));
+  c10_cuda_check_implementation(__FILE__, __FUNCTION__, __LINE__, false);
+
+  C10_CUDA_ERROR_HANDLED(cudaMemAdvise(
+      uvm_assertions_ptr,
+      sizeof(DeviceAssertionsData),
+      cudaMemAdviseSetPreferredLocation,
+      cudaCpuDeviceId));
+  c10_cuda_check_implementation(__FILE__, __FUNCTION__, __LINE__, false);
+
+  // GPU will establish direct mapping of data in CPU memory, no page faults
+  // will be generated
+  C10_CUDA_ERROR_HANDLED(cudaMemAdvise(
+      uvm_assertions_ptr,
+      sizeof(DeviceAssertionsData),
+      cudaMemAdviseSetAccessedBy,
+      0));
+  c10_cuda_check_implementation(__FILE__, __FUNCTION__, __LINE__, false);
+
+  // Initialize the memory from the CPU; otherwise, pages may have to be created
+  // on demand. We think that UVM documentation indicates that first access may
+  // not honor preferred location, which would be bad, if true, because we want
+  // this memory on the host so we can access it post-assertion. Initializing
+  // this on the CPU helps ensure that that's where the memory will live.
+  *uvm_assertions_ptr = DeviceAssertionsData();
+
+  // Ownership and lifetime management of `uvm_assertions_ptr` now passes to the
+  // uvm_assertions unique_ptr vector
+  uvm_assertions.at(device_num).reset(uvm_assertions_ptr);
+#endif
+
+  return uvm_assertions_ptr;
+}
+
+int CUDAKernelLaunchRegistry::get_device_count() {
+  int device_count = -1;
+  C10_CUDA_ERROR_HANDLED(cudaGetDeviceCount(&device_count));
+  c10_cuda_check_implementation(__FILE__, __FUNCTION__, __LINE__, false);
+  return device_count;
```

Comment:
It's not clear why this has to be a method on the class

Reply:
can't you call `c10::cuda::device_count/current_device()` directly? There's a circular dependency between the headers, but since in .cpp file it should be ok.

I'd avoid duplicating the calls.

Reply:
Sure, but it still doesn't have to be a method on the class :P